File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Delete_the_Middle_Node_of_a_Linked_List/kihyun.c
#include <stdio.h>
#include <string.h>
#include <limits.h>
#include <math.h>
#include <stdlib.h>

/**
 * Definition for singly-linked list.
 */

struct ListNode {
     int val;
     struct ListNode *next;
};
 
struct ListNode* deleteMiddle(struct ListNode* head) {
    struct ListNode* mid = head;
    struct ListNode* before_mid = head;
    struct ListNode* end = head;
    if (head == NULL || head->next == NULL){
        return NULL;
    }

    while(end->next != NULL){
        before_mid = mid;
        mid = mid->next;
        end = end->next->next;
        if (end == NULL) break;
    }
    before_mid->next = mid->next;
    return head; 
}

File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Delete_the_Middle_Node_of_a_Linked_List/kihyun.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        mid = head
        end = head
        before = mid
        if head.next == None:
            return None
        while (end.next != None):
            if end.next.next != None:
                end = end.next.next
            else:
                end = end.next
            before = mid
            mid = mid.next
        before.next = mid.next
        return head
        